Assets/__Scripts/MapEditor/Audio/AudioManager.cs

```diff
@@ -1,96 +1,180 @@
+        Debug.Log("WaveformGenerator: Starting Background Thread");
 
+        Thread bgThread = new Thread(preProcessData);
+        backgroundThreads.Add(bgThread);
+        bgThread.Start();
+    }
+
+    // This will be triggered if the game wants to close or we're exiting to the main menu
+    // Remove all chunks from the queue so the background threads quit
+    void OnDestroy()
+    {
+        int dummy;
+        while (chunkQueue.TryDequeue(out dummy)) { }
     }
 
-    public void PopulateData()
+    public void preProcessData()
     {
-        // copying last values level "up"
-        for (int i = ColumnsPerChunk - 1; i > 0; i--)
-            Array.Copy(_bandVolumes[i - 1], _bandVolumes[i], _bands.Count - 1);
+        waveformData.chunks = (int) Math.Ceiling(clipLength / secondPerChunk);
+        preProcessedSamples = new float[this.numTotalSamples];
 
-        // reading current samples
-        float[] samples = new float[SAMPLE_COUNT];
-        _audioSource.GetSpectrumData(samples, 0, FFTWindow.BlackmanHarris);
+        // Average all audio channels together
+        for (int i = 0; i < multiChannelSamples.Length; i++)
+        {
+            preProcessedSamples[i / numChannels] = multiChannelSamples[i] / numChannels;
+        }
 
-        //Smoothing code largely taken from beat saber but also simplified a lot
-        float deltaTime = Time.deltaTime;
-        for (int i = 0; i < SAMPLE_COUNT; i++)
+        Debug.Log("WaveformGenerator: Combine Channels done " + preProcessedSamples.Length);
+
+        for (int i = waveformData.chunksLoaded; i < waveformData.chunks; i++)
         {
-            float num = samples[i];
-            processedSamples[i] = Mathf.Lerp(processedSamples[i], num, deltaTime * 128f);
+            chunkQueue.Enqueue(i);
         }
 
-        float[] bandVolumes = new float[_bands.Count - 1];
-        for (int i = 1; i < _bands.Count; i++)
+        // How many audio samples wide a column is, we will likely use more samples
+        // than the width to perform the FFT which has a smoothing effect
+        sampleOffset = ((secondPerChunk / (float)ColumnsPerChunk) * sampleRate);
+
+        // Precalculate the window function for our sample size
+        windowCoefs = DSP.Window.Coefficients(DSP.Window.Type.BH92, (uint)SAMPLE_COUNT);
+        scaleFactor = DSP.Window.ScaleFactor.Signal(windowCoefs);
+
+        hzStep = sampleRate / (float)SAMPLE_COUNT;
+        waveformData.initBandVolumes(ColumnsPerChunk);
+
+        // Try and leave at least one cpu core for the main thread, limited by MAX_THREADS
+        int threadCount = Math.Max(1, Math.Min(Environment.ProcessorCount - 1, MAX_THREADS));
+        for (int i = 0; i < threadCount; i++)
+        {
+            Thread FFTThread = new Thread(performFFTThreaded);
+            backgroundThreads.Add(FFTThread);
+            FFTThread.Start();
+        }
+
+        Debug.Log("Background Thread Completed");
+    }
+
+    public void performFFTThreaded()
```

Comment:
This looks good, but my biggest issue here is the inconsistent naming conventions. I'd appreciate it if methods and public members use `PascalCase`. This also applies to `WaveformCache.cs`.